### ISSUE TITLE
Lab 9 peer review by Martini Matteo s314786

### ISSUE THREAD

**Issue by MatteMartini** [FedeBucce/Computational_intelligence#4]
The code is well-written, and the comments in it really help with understanding. I like that you chose to implement not just an evolutionary algorithm but also an island model, using some techniques we've learned about recently, like extinction and migrants.

A suggestion I have is to try using slightly different settings for your algorithms. In my opinion, using 2000 iterations with a population size of 10 and offspring size of 60 seems a bit unbalanced. You can get a much higher max_fitness by increasing your population and offspring while reducing the number of iterations to avoid too many fitness calls. 
For example, with 1500 iterations and a population size of 200, offspring size of 300, you can achieve 100% max_fitness for PROBLEM_SIZE=1 in both models you used and get close to 100% for PROBLEM_SIZE=2. In any case, performance improves considerably for all PROBLEM_SIZE, with a more extensive selection among individuals in your population, resulting in more widespread mutations and crossovers. 

![Screenshot 2023-12-05 132409](https://github.com/FedeBucce/Computational_intelligence/assets/127160498/c1518e96-6540-4cf8-a5e9-211ab3009c49)

Either way, your work is really good—congratulations!

**Comment by s320168** [comment]
The solution implements a good Genetic Algorithm with both a classical population and a island model. The first one follows the literature given by the course, quickly returning its results. Personally I would have used a OFFSPRING_SIZE larger than the POPULATION_SIZE and bigger numbers in general for the two variables (to get closer to the solution while losing a bit of speed). The island model variant manages to get slightly better results (with POPULATION_SIZE=200 and OFFSPRING_SIZE=300 it sometimes can complete problems with size 1). The extinction every 50 generations is a nice addition to escape from a plateau of fitness values. One thing that I noticed is that when migrations or extinctions happen the niches should get ordered before extracting the best individual of the current generation (this error is noticeable as a drop in the fitness graphs).
In conclusion this is a very nice implementation, it just needs a bit of tuning.

**Comment by FedeBucce** [comment]
Thank you for your review!
